### ISSUE TITLE
Reversed ListView range is not maintained when using RangeMaintainingScrollPhysics

### ISSUE THREAD

**Issue by VijayakumarMariappan** [flutter/flutter#155152]
### Steps to reproduce

- Run the provided sample.
- Scroll to the top and notice that the range in the list view has changed to the size of the recently added list items, rather than maintaining the same range. For example, scroll to the top, where you see 'Item 0,' and wait for the next item to be added to the list view. When the new item is added, the list view automatically scrolls down to the size of newly added item, resulting in a change in the range.

### Expected results

As we are using `RangeMaintainingScrollPhysics `in the list view, the list should stay stable while scrolling, even when items are added dynamically.

### Actual results

The list view automatically scrolled down to the size of the newly added item.

![List view range maintaining issue](https://github.com/user-attachments/assets/09fce921-4fa6-4a34-a078-6c720c0fe4bb)

### Code sample

```dart
import 'dart:async';

import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatefulWidget {
  const MainApp({super.key});

  @override
  State<MainApp> createState() => _MainAppState();
}

class _MainAppState extends State<MainApp> {
  late List<String> _baseItems;
  final List<String> _loadingItems = <String>[];

  @override
  void initState() {
    _baseItems = List<String>.generate(10000, (index) => 'Item $index');
    Timer.periodic(const Duration(seconds: 1), (timer) {
      if (_baseItems.isEmpty) {
        timer.cancel();
        return;
      }

      setState(() {
        _loadingItems.add(_baseItems.removeAt(0));
      });
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: ListView.builder(
          reverse: true,
          physics: const RangeMaintainingScrollPhysics(),
          itemCount: _loadingItems.length,
          itemBuilder: (BuildContext context, int index) {
            return Padding(
              padding: const EdgeInsets.all(5.0),
              child: ListTile(
                title: Text('Item ${_loadingItems.length - index - 1}'),
                tileColor: Colors.blue[100],
              ),
            );
          },
        ),
      ),
    );
  }

  @override
  void dispose() {
    _baseItems.clear();
    _loadingItems.clear();
    super.dispose();
  }
}
```

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

```console
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.24.1, on Microsoft Windows [Version 10.0.22621.4169], locale en-US)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[!] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/to/windows-android-setup for more details.
[√] Chrome - develop for the web
[√] Visual Studio - develop Windows apps (Visual Studio Enterprise 2022 17.11.3)
[!] Android Studio (not installed)
[√] VS Code (version 1.93.0)
[√] Connected device (3 available)
[√] Network resources

! Doctor found issues in 2 categories.
```

**Comment by danagbemava-nc** [comment]
Reproducible using the code sample & steps provided above. 

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.24.3, on macOS 14.6.1 23G93 darwin-arm64, locale en-US)
    • Flutter version 3.24.3 on channel stable at /Users/deanli/dev/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (2 days ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3
```

```
[!] Flutter (Channel master, 3.26.0-1.0.pre.113, on macOS 14.6.1 23G93 darwin-arm64, locale en-US)
    • Flutter version 3.26.0-1.0.pre.113 on channel master at /Users/deanli/dev/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 1d5f36f014 (8 hours ago), 2024-09-12 22:59:31 -0400
    • Engine revision 04802b7790
    • Dart version 3.6.0 (build 3.6.0-248.0.dev)
    • DevTools version 2.40.0-dev.1
```
</details>

**Comment by Piinks** [comment]
This appears to be working as expected, since it is reversed, the bottom of the view is the zero position. As items are inserted, it maintains the zero position.

**Comment by VijayakumarMariappan** [comment]
@Piinks I have a reversed list with 30 items and am adding items one by one. After adding 20 items, 10 are visible, and the other 10 have moved outside the visible range. I then scroll up to a range where the visible items are 5-15.
I expect the visible range to remain consistent when new items are added, due to `RangeMaintainingScrollPhysics`. However, the visible range shifts, and instead of staying at 5-15, it changes to 6-16 when a new item is added. This behavior seems incorrect to me.
If the range isn't maintained in this case, it defeats the purpose of using `RangeMaintainingScrollPhysics` in a reversed `ListView`. The expectation is that the visible range should stay fixed as items are added, but this isn't happening, making the feature ineffective for reversed lists.

**Comment by Piinks** [comment]
> If the range isn't maintained in this case

It sounds like there may be a misconception about how RangeMaintainingScrollPhysics are supposed to work. The range is specified by the content, and so adding items changes the range.
Can you share more about the expected behavior? I am not sure I understand form the last comment.

**Comment by VijayakumarMariappan** [comment]
@Piinks I will continue from the previous example where I have a reversed list containing 30 items. I am adding list items one by one dynamically with a delay of 1 second. Let's consider that there are 10 list items displayed in the visible area. After adding 20 items, 10 items are visible, while the other 10 have moved out of the visible range at the top. 

When I scroll up to view items 5 through 15, new list items are still being added to the list view every second.

> The range is specified by the content, and so adding items changes the range.

At this point, the visible list items are now 5 to 15. When a new list item is added, the overall range should change from 20 to 21, and we expect the visible list items to remain stable at 5 to 15. However, this is not the case; it changes to 6 to 16.

**Comment by chunhtai** [comment]
I also don't think this is a bug because the scroll position is maintained. I think we need new API to achieve the goal you want

**Comment by gnprice** [comment]
This is correct behavior. It isn't really about `reverse: true`, nor about RangeMaintainingScrollPhysics. The reason you're seeing this behavior where the visible item numbers change is this line in the example code:
```dart
                title: Text('Item ${_loadingItems.length - index - 1}'),
```

So when for example there are 30 items in the list (meaning `_loadingItems.length` is 30), and you see an item with the text "Item 5", that isn't the item at index 5 in the list. That's the item where `_loadingItems.length - index - 1` is 5, which means `index` is 24. Similarly the text "Item 15" is on the item at index 14.

The list view isn't inspecting the text that you put in the widgets, and doesn't know anything about that. What it knows is the indexes of the items: it knows that the item at index 14 is at the bottom (the start) of the list, and the item at index 24 is at the top (the end).

When you add another item to `_loadingItems`, the ListView does its job and continues to show the item with index 14 at the bottom and the item with index 24 at the top. Because the widget you put inside the list items says `Text('Item ${_loadingItems.length - index - 1}')`, and because the length of the list is now 31, that means the items that say "Item 16" and "Item 6" respectively.

The list view is still showing you the same items, as designed. It's the content of those items that has changed, because the code you wrote for them depends on something that changed, namely the length of the list.

---

If you want the list to remain scrolled to a constant distance from the top of the list, instead of the bottom of the list, you can do that by leaving `reverse` false so that the list starts at the top.

If you want the list to continue showing "the same" items in the viewport both when items are added/removed at the start of the list, and when they're added/removed at the end, then `ListView.builder` will not do the job. As @chunhtai says above, supporting that out of the box would require new API.

I think you may be able to do it yourself using `ListView.custom`, with your own `SliverChildDelegate` subclass, but that will take some work.

Zulip is a messaging app written in Flutter, and has a version of this issue. You might find the discussion there informative:
* https://github.com/zulip/zulip-flutter/issues/83

For us at Zulip, it's not a high-priority issue so I don't plan to build a solution anytime soon. Depending on how much it's bugging you in your app, you might find this part helpful (which we're doing soon in Zulip for other reasons anyway):
> A good 90% solution here would be to build the message list out of two list slivers facing back to back, rather than one list sliver facing upward. This is a standard Flutter solution for having a list that grows in both directions; see [example in the CustomScrollView doc](https://api.flutter.dev/flutter/widgets/CustomScrollView-class.html#widgets.CustomScrollView.2).
